src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -61,18 +62,66 @@ public void stop() {
     stops_.set(true);
   }
 
+  /**
+   * Store result of current rule execution
+   * @param result
+   */
+  public void putResult(Object result) {
+    results_.put(currentRule_.getId(), result);
+  }
+
+  public Object getResult(long ruleId) {
+    return results_.get(ruleId);
+  }
+
+  private void applyRules(List<Rule> rules, FObject obj, FObject oldObj) {
+    for (Rule rule : rules) {
+      if ( stops_.get() ) return;
+
+      currentRule_ = rule;
+      if ( rule.f(getX(), obj, oldObj)
+        && rule.getAction() != null
+      ) {
+        rule.apply(getX(), obj, oldObj, this);
+        saveHistory(rule, obj);
+      }
+    }
+  }
+
   private void asyncApplyRules(List<Rule> rules, FObject obj, FObject oldObj) {
     ((FixedThreadPool) getX().get("threadPool")).submit(getX(), new ContextAgent() {
       @Override
       public void execute(X x) {
         for (Rule rule : rules) {
           if ( stops_.get() ) return;
 
-          if ( rule.f(getX(), obj, oldObj) ) {
+          currentRule_ = rule;
+          if ( rule.f(getX(), obj, oldObj)
+            && rule.getAsyncAction() != null
+          ) {
             rule.asyncApply(getX(), obj, oldObj, RuleEngine.this);
+            saveHistory(rule, obj);
```

Comment:
saveHistory saves the result of the rule, but with async, it's not complete yet.  Believe apply should saveHistory on completion.

Reply:
This is inside the thread pool and it would complete as is.